### PR TITLE
KerasCV Object Detection API Overview

### PR DESCRIPTION
**Pull request by LukeWood**

### CODE REVIEW

**Comment by LukeWood** [comment]
A thought I just had was to make the monitor be on Val MAP instead of loss.

File: guides/keras_cv/retina_net_overview.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
+"""
+Title: Train an Object Detection Model on Pascal VOC 2007 using KerasCV
+Author: [lukewood](https://lukewood.xyz)
+Date created: 2022/08/02
```
Comment (fchollet) [inline guides/keras_cv/retina_net_overview.py:4]: Nit: update dates

File: guides/keras_cv/retina_net_overview.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+Last modified: 2022/08/02
+Description: Use KerasCV to train a RetinaNet on Pascal VOC 2007.
+"""
+
+"""
+## Overview
+
+KerasCV offers a complete set of APIs to allow you to train your own state-of-the-art,
```
Comment (fchollet) [inline guides/keras_cv/retina_net_overview.py:12]: Nit: cut out "to allow you", simpler is better

File: guides/keras_cv/retina_net_overview.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+data augmentation techniques, models, and COCO metrics.
+
+To get started, let's sort out all of our imports and define global configuration parameters.
+"""
+
+import matplotlib.pyplot as plt
+import tensorflow as tf
+import tensorflow_datasets as tfds
+from tensorflow import keras
+from tensorflow.keras import callbacks as callbacks_lib
```
Comment (fchollet) [inline guides/keras_cv/retina_net_overview.py:23]: This is only used 3x in 1 location, I'd recommend just accessing `keras.callbacks` instead of importing the submodule

File: guides/keras_cv/retina_net_overview.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+import keras_cv
+from keras_cv import bounding_box
+
+BATCH_SIZE = 8
+EPOCHS = 1
+
+"""
+## Data loading
+
+In this guide, we use the data-loading function: `keras_cv.loaders.pascal_voc.load()`.
```
Comment (fchollet) [inline guides/keras_cv/retina_net_overview.py:35]: Why not use `keras_cv.datasets` for dataset loaders (consistent with `keras.datasets` for users already familiar with it)? "loader" is a very generic term, to be avoided (it could refer to anything)

File: guides/keras_cv/retina_net_overview.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    split="train", bounding_box_format="xywh", batch_size=9
+)
+
+
+def visualize_dataset(dataset, bounding_box_format):
+    color = tf.constant(((255.0, 0, 0),))
+    plt.figure(figsize=(10, 10))
+    iterator = iter(dataset)
+    for i in range(9):
+        example = next(iterator)
```
Comment (fchollet) [inline guides/keras_cv/retina_net_overview.py:68]: Instead of these three lines, consider:

```
for i, example in enumerate(dataset.take(9)):
```

File: guides/keras_cv/retina_net_overview.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+train_ds, train_dataset_info = keras_cv.loaders.pascal_voc.load(
+    bounding_box_format="xywh", split="train", batch_size=BATCH_SIZE
+)
+val_ds, val_dataset_info = keras_cv.loaders.pascal_voc.load(
+    bounding_box_format="xywh", split="validation", batch_size=BATCH_SIZE
+)
+
+augmentation_layers = [
+    # keras_cv.layers.RandomShear(x_factor=0.1, bounding_box_format='xywh'),
+    # TODO(lukewood): add color jitter and others
```
Comment (fchollet) [inline guides/keras_cv/retina_net_overview.py:112]: We need to remember to update this before publishing

File: guides/keras_cv/retina_net_overview.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+"""
+Great!  We now have a bounding box friendly augmentation pipeline.
+
+Next, let's unpackage our inputs from the preprocessing dictionary, and prepare to feed
+the inputs into our model.
+"""
+
+
+def unpackage_dict(inputs):
```
Comment (fchollet) [inline guides/keras_cv/retina_net_overview.py:133]: `unpack`? `dict_to_tuple`?

File: guides/keras_cv/retina_net_overview.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+freeze the backbone before training.  When `include_rescaling=True` is set, inputs to
+the model are expected to be in the range `[0, 255]`.
+"""
+
+model = keras_cv.models.RetinaNet(
+    classes=20,
+    bounding_box_format="xywh",
+    backbone="resnet50",
+    backbone_weights="imagenet",
+    include_rescaling=True,
```
Comment (fchollet) [inline guides/keras_cv/retina_net_overview.py:161]: Please provide quick in-line comments for each of these arguments, as this code block will likely be the first contact with this API for many readers

File: guides/keras_cv/retina_net_overview.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    callbacks_lib.TensorBoard(log_dir="logs"),
+    callbacks_lib.EarlyStopping(patience=50),
+    callbacks_lib.ReduceLROnPlateau(patience=20),
+]
+
+"""
+And run `model.fit()`!
+"""
+
+model.fit(
```
Comment (fchollet) [inline guides/keras_cv/retina_net_overview.py:251]: We should add an inference section as well, including visualizing inference results (we could just drop saved images of the final results in the markdown since the model here is not trained to convergence)

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! Looking great!

File: guides/keras_cv/retina_net_overview.py (hunk 4 above)
Comment (LukeWood) [reply guides/keras_cv/retina_net_overview.py:35]: https://github.com/keras-team/keras-cv/pull/711/files

File: guides/keras_cv/retina_net_overview.py (hunk 7 above)
Comment (LukeWood) [reply guides/keras_cv/retina_net_overview.py:133]: Done

File: guides/keras_cv/retina_net_overview.py (hunk 9 above)
Comment (LukeWood) [reply guides/keras_cv/retina_net_overview.py:251]: Sounds good, once we finish up stabilizing the API I'll run this on Colab and generate some nice visuals.

File: guides/keras_cv/retina_net_overview.py [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+training targets are used by `keras_cv.losses.ObjectDetectionLoss()` to train the
+network.
+"""
+
+"""
+## Optimizer
+
+For training, we use a SGD optimizer with a piece-wise learning rate schedule
+consisting of a warm up followed by a ramp up, then a ramp.
+Below, we construct this using a `keras.optimizers.schedules.PiecewiseConstantDecay`
```
Comment (fchollet) [inline guides/keras_cv/retina_net_overview.py:186]: I had missed this -- the text does not match the code. The code does not currently use a schedule.

**Review by fchollet — APPROVED** [review]
LGTM, thanks for the updates! I applied some copyedits, please generate the files after pulling them.

**Comment by LukeWood** [comment]
@fchollet updated to use the new API w/ evaluate() & predict

This is TPU compatible, etc

File: guides/keras_cv/retina_net_overview.py [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+KerasCV supports a `bounding_box_format` argument in all components that process
+bounding boxes.  To match the KerasCV API style, it is recommended that when writing a
+custom data loader, you also support a `bounding_box_format` argument.
+This makes it clear to those invoking your data loader what format the bounding boxes
+are in.
+
+For example:
+
+```python
+train_ds, ds_info = keras_cv.datasets.pascal_voc.load(split='train', bounding_box_format='xywh', batch_size=8)
```
Comment (fchollet) [inline guides/keras_cv/retina_net_overview.py:44]: Shorten line

```
dataset, dataset_info = keras_cv.datasets.pascal_voc.load(
    split="train", bounding_box_format="xywh", batch_size=8
)
```

File: guides/keras_cv/retina_net_overview.py [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+
+Our data comesloaded into the format
+`{"images": images, "bounding_boxes": bounding_boxes}`.  This format is supported in all
+KerasCV preprocessing components.
+
+Let's load some data and verify that our data looks as we expect it to.
+"""
+
+dataset, dataset_info = keras_cv.datasets.pascal_voc.load(
+    split="train", bounding_box_format="xywh", batch_size=9
```
Comment (fchollet) [inline guides/keras_cv/retina_net_overview.py:58]: Why 9 rather than 8? A priori 8 will lead to better GPU memory management

File: templates/keras_cv/index.md [hunk 13]
```diff
@@ -35,32 +35,28 @@ import tensorflow as tf
 from tensorflow import keras
 import tensorflow_datasets as tfds
 
-preprocessing_layers = [
-    keras_cv.layers.RandomResizedCrop(
```
Comment (fchollet) [inline templates/keras_cv/index.md:39]: Why is this layer dropped?

File: guides/keras_cv/retina_net_overview.py (hunk 12 above)
Comment (LukeWood) [reply guides/keras_cv/retina_net_overview.py:58]: Ah I see, I was trying to support the visualization at the end which requires 9 samples

File: templates/keras_cv/index.md (hunk 13 above)
Comment (LukeWood) [reply templates/keras_cv/index.md:39]: we're changing the API soon and I don't want to push the layer anymore until the API is done.

**Review by fchollet — APPROVED** [review]
LGTM, thanks -- you can add the generated files.

**Comment by LukeWood** [comment]
ok @fchollet I fixed the links!

**Comment by LukeWood** [comment]
@fchollet PTAL!  GenFiles added

File: guides/keras_cv/retina_net_overview.py [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+def visualize_dataset(dataset, bounding_box_format):
+    color = tf.constant(((255.0, 0, 0),))
+    plt.figure(figsize=(10, 10))
+    for i, example in enumerate(dataset.take(9)):
+        images, boxes = example["images"], example["bounding_boxes"]
+        boxes = keras_cv.bounding_box.convert_format(
+            boxes, source=bounding_box_format, target="rel_yxyx", images=images
+        )
+        boxes = boxes.to_tensor(default_value=-1)
+        plotted_images = tf.image.draw_bounding_boxes(images, boxes[..., :4], color)
```
Comment (fchollet) [inline guides/keras_cv/retina_net_overview.py:76]: `draw_bounding_boxes` doesn't have a way to control line thickness, does it? I found the boxes a bit hard to parse visually.

**Review by fchollet — APPROVED** [review]
LGTM, thank you! Great work on this guide.

File: guides/keras_cv/retina_net_overview.py (hunk 14 above)
Comment (LukeWood) [reply guides/keras_cv/retina_net_overview.py:76]: it does not :/

**Comment by LukeWood** [comment]
Can’t wait to see this live!  Thanks for the reviews!
